src/python/pants/backend/shell/shell_command.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from textwrap import dedent
+
+from pants.backend.shell.target_types import (
+    ShellCommandCommandField,
+    ShellCommandLogOutputField,
+    ShellCommandOutputsField,
+    ShellCommandSources,
+    ShellCommandToolsField,
+)
+from pants.core.target_types import FilesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, Directory, MergeDigests, Snapshot
+from pants.engine.process import (
+    BashBinary,
+    BinaryNotFoundError,
+    BinaryPathRequest,
+    BinaryPaths,
+    Process,
+    ProcessResult,
+    SearchPath,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    Sources,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+class GenerateFilesFromShellCommandRequest(GenerateSourcesRequest):
+    input = ShellCommandSources
+    output = FilesSources
```

Comment:
To respond to the earlier thread: I think this is the right choice for now to start, rather than `Sources`.

Note also that callers must set `enable_codegen=True` for this codegen to happen.